OutOfSchool/OutOfSchool.Redis/MultiLayerCache.cs

```diff
@@ -0,0 +1,125 @@
+﻿using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Options;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OutOfSchool.Redis;
+
+public sealed class MultiLayerCache : IMultiLayerCacheService, IDisposable
+{
+    private readonly ReaderWriterLockSlim cacheLock = new ReaderWriterLockSlim();
+    private readonly IMemoryCache _memoryCache;
+    private readonly ICacheService _cacheService;
+    private readonly MemoryCacheConfig memoryCacheConfig;
+
+    public MultiLayerCache(
+        ICacheService externalCache,
+        IMemoryCache inMemoryCache,
+        IOptions<MemoryCacheConfig> memoryCacheConfig
+    )
+    {
+        this._memoryCache = inMemoryCache;
+        this._cacheService = externalCache;
+        this.memoryCacheConfig = memoryCacheConfig.Value;
+    }
+
+    public Task ClearCacheAsync(string key)
+    {
+        return ExecuteMultipleAsync(() =>
+        {
+            cacheLock.EnterWriteLock();
+            try
+            {
+                _memoryCache.Remove(key);
+                _cacheService.ClearCacheAsync(key);
+            }
+            finally
+            {
+                cacheLock.ExitWriteLock();
+            }
+        });
+    }
+
+    public async Task<T> GetOrAddAsync<T>(string key, Func<Task<T>> newValueFactory, TimeSpan? absoluteExpirationRelativeToNowInterval = null, TimeSpan? slidingExpirationInterval = null)
+    {
+        T returnValue = default;
+
+        await ExecuteMultipleAsync(() =>
+        {
+            cacheLock.EnterReadLock();
+            
+            try
+            {
+                if (_memoryCache.TryGetValue(key, out T value))
+                {
+                    returnValue = value;
+                }
+            }
+            finally
+            {
+                cacheLock.ExitReadLock();
+            }
+        });
+
+        if (EqualityComparer<T>.Default.Equals(returnValue, default))
+        {
+            returnValue = await _cacheService.GetOrAddAsync(key, newValueFactory, absoluteExpirationRelativeToNowInterval, slidingExpirationInterval);
+            await SetAsync(key, returnValue, absoluteExpirationRelativeToNowInterval, slidingExpirationInterval);
+        }
+
+        return returnValue;
+    }
+
+    public Task RefreshAsync(string key)
```

Comment:
As a suggestion
```
public Task RefreshAsync(string key)
{
    cacheLock.EnterWriteLock();
    return _cacheService.RefreshAsync(key).ContinueWith(_ => cacheLock.ExitWriteLock());
}
```
This way it removes the `ExecuteMultipleAsync` and really awaits the `_cacheService.RefreshAsync()`

Reply:
Looks pretty.

Reply:
It would be cool if `CacheService` `RefreshAsync()` be written in this way. It is not.

Reply:
Deleted Refresh and from the interface